docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,176 @@
+# Grafana Agent backwards compatibility 
+
+* Date: 2023-05-25
+* Author: Robert Fratto (@rfratto)
+* PR: [grafana/agent#3981](https://github.com/grafana/agent/pull/3981)
+* Status: Draft
+
+Grafana Agent has been following [semantic versioning](https://semver.org/)
+since its inception. After three years of development and 33 minor releases,
+the project is on trajectory to have a 1.0 release. 
+
+In the context of semantic versioning, a 1.0 release indicates that future
+minor releases have backwards compatibility with older minor releases in the
+same major release; version 1.1 is backwards compatible with version 1.0.
+Having major releases signals to users when they can upgrade with more
+confidence (minor releases) and when an upgrade may take more time (new major
+releases). 
+
+However, Grafana Agent is a large project, with a large surface area for what
+may be considered part of the backwards compatibility guarantees. This proposal
+formally establishes what parts of Grafana Agent will be protected by backwards
+compatibility. 
+
+## Goals 
+
+- Set expectations for what is covered by backwards compatibility. 
+- Set expectations for when upgrades to new major versions will be forced.
+
+## Proposal 
+
+Backwards compatibility means that a user can upgrade their version of Grafana
+Agent without needing to make any changes with the way they interact with
+Grafana Agent. 
+
+### Scope of backwards compatibility   
```

Comment:
The main thing which bothers me is whether backwards compatibility guarantees are more effort that it's worth? I wish there was more information about this in the RFC. For example:
* Is there anecdotal evidence from users that they are struggling to upgrade? Or that there could be changes which would prevent them from upgrading?
* Are any user adoption metrics showing that adoption of new versions is lagging behind, and this is something this RFC could help with? E.g. did we notice that in the past when we introduced a big breaking change, adoption of the new version was not as quick as usual?
* Even if there are pain points which compatibility could help with, if we simply list them in the exception category, then our compatibility guarantees still wouldn't help the users. Knowing more about the problem we're solving could help us determine if our exceptions list is too broad.

Reply:
I'd like to clarify that this proposal is less about guaranteeing backwards compatibility, and more to scope what backwards compatibility means given that we follow semantic versioning. 

There are a few things that are leading to this:

* We've been asked in the past to release 1.0, since 1.0 gives the impression of stability and maturity. 
* We have always followed semantic versioning, where versions >0.X must be backwards compatible with the same major release version. 
* So, to release 1.0, we need to define the scope of what is in our public API to know when a release 2.0 would be required. 

To your questions: 

We do see in our usage stats that more than half of reporting installations are several versions behind; only about 7% of users are on the latest v0.33 release. 

In my experience, most users typically don't upgrade unless they need something new or need a specific fix. This makes sense, since when the agent is acting perfectly you leave it in the background and don't think about it. This does mean that when users _do_ need to upgrade, they're likely several versions behind. 

Ignoring backwards compatibility guarantees can lead to accumulating a growing backlog of work to do when a user _does_ need to finally upgrade. My concern is that this is a bad user experience: we're either asking them to stay on top of updating software they don't need to update just for the sake of staying as compatible as possible, or risk being blocked on functionality they need by going through a growing checklist of breaking change items to address.

I think we've been told once or twice to not make as many breaking changes as we have, but disregarding that, I'm mainly proposing we build software that I personally would want to use: clear expectations about breaking changes so that most upgrades are as pain-free as possible.